NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,474 @@
+#include "pch.h"
+#include <filesystem>
+#include <sstream>
+#include <fstream>
+#include "squirrel/squirrel.h"
+#include "util/utils.h"
+#include "mods/modmanager.h"
+#include "modsavefiles.h"
+#include "rapidjson/document.h"
+#include "rapidjson/writer.h"
+#include "rapidjson/stringbuffer.h"
+#include "config/profile.h"
+#include "core/tier0.h"
+#include "rapidjson/error/en.h"
+
+SaveFileManager* g_saveFileManager;
+int MAX_FOLDER_SIZE = 52428800; // 50MB (50 * 1024 * 1024)
+fs::path savePath;
+
+uintmax_t GetSizeOfFolderContentsMinusFile(fs::path dir, std::string file)
+{
+	uintmax_t result = 0;
+	for (const auto& entry : fs::directory_iterator(dir))
+	{
+		if (entry.path().filename() == file)
+			continue;
+		result += fs::file_size(entry.path());
+	}
+	return result;
+}
+
+template <ScriptContext context> void SaveFileManager::SaveFileAsync(fs::path file, std::string contents)
+{
+	auto mutex = std::ref(mutexMap[file]);
+	std::thread writeThread(
+		[mutex, file, contents]()
+		{
+			try
+			{
+				mutex.get().lock();
+
+				fs::path dir = file.parent_path();
+				// this actually allows mods to go over the limit, but not by much
+				// the limit is to prevent mods from taking gigabytes of space,
+				// this ain't a cloud service.
+				if (GetSizeOfFolderContentsMinusFile(dir, file.filename().string()) + contents.length() > MAX_FOLDER_SIZE)
+				{
+					// tbh, you're either trying to fill the hard drive or use so much data, you SHOULD be congratulated.
+					spdlog::error(fmt::format("Mod spamming save requests? Folder limit bypassed despite previous checks. Not saving."));
+					mutex.get().unlock();
+					return;
+				}
+
+				std::ofstream fileStr(file);
+				if (fileStr.fail())
+				{
+					mutex.get().unlock();
+					return;
+				}
+
+				fileStr.write(contents.c_str(), contents.length());
+				fileStr.close();
+
+				mutex.get().unlock();
+				// side-note: this causes a leak?
+				// when a file is added to the map, it's never removed.
+				// no idea how to fix this - because we have no way to check if there are other threads waiting to use this file(?)
+				// tried to use m.try_lock(), but it's unreliable, it seems.
+			}
+			catch (std::exception ex)
+			{
+				spdlog::error("SAVE FAILED!");
+				mutex.get().unlock();
+				spdlog::error(ex.what());
+			}
+		});
+
+	writeThread.detach();
+}
+
+template <ScriptContext context> int SaveFileManager::LoadFileAsync(fs::path file)
+{
+	int handle = ++m_iLastRequestHandle;
+	auto mutex = std::ref(mutexMap[file]);
+	std::thread readThread(
+		[mutex, file, handle]()
+		{
+			try
+			{
+				mutex.get().lock();
+
+				std::ifstream fileStr(file);
+				if (fileStr.fail())
+				{
+					spdlog::error("A file was supposed to be loaded but we can't access it?!");
+					// we should throw a script error here. But idk how lmao
+					g_pSquirrel<context>->AsyncCall("NSHandleLoadResult", handle, "");
```

Comment:
added optional onfailure callback if the read fails.

calls onsuccess with empty string/table if not used and read fails otherwise (for those who'd rather that)

Reply:
just making a note that this is handled in script now